### DISCUSSION TITLE
Massive lags in song select

### DISCUSSION THREAD

**Discussion by nyannoying1337** [ppy/osu#15160] [Q&A]
When i start scrolling through song select osu! becomes really laggy. its fine once pictures are loaded but they keep on reloading every exit from song select / restart of the game 

Video: https://streamable.com/uzgvqu
Logs:
[performance-draw.log](https://github.com/ppy/osu/files/7360644/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/7360645/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/7360646/performance-update.log)
[performance.log](https://github.com/ppy/osu/files/7360647/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/7360648/performance-audio.log)
 
 System specs: Windows 11 x64, i5-10400f, 16GB RAM, Nvidia GTX 1050 Ti

**Comment by smoogipoo** [comment]
Not sure right off the bat, it looks to be texture upload related judging by the logs. Can you check whether you have the latest graphics drivers installed?

**Reply by nyannoying1337** [reply]
@smoogipoo i actually wasn't. so i upgraded from Game Ready Driver 472.12 to Game Ready Driver 496.13.
The issue still persists. Sometimes when i restart the game, the maps i **previously** loaded are fine as example here https://streamable.com/7fi3ql but sometimes it doesn't and it starts to lag regardless, this is however **always** the case after a pc restart meaning it would lag with every single scroll i do

( the logs are from after the pc restart )

[performance.log](https://github.com/ppy/osu/files/7361436/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/7361437/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/7361438/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/7361439/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/7361440/performance-update.log)

**Reply by peppy** [reply]
Probably hitting the memory limit of the GPU at some point and swapping out to RAM.

**Reply by nyannoying1337** [reply]
![image](https://user-images.githubusercontent.com/17437302/137661553-002946cb-52a4-4b02-97bc-04eebaaa493a.png)
would this give you guys any more clues?

**Reply by smoogipoo** [reply]
I remember a similar issue when we upgraded to EFCore 5, so perhaps another cause could be the number of beatmaps loaded?

Try opening the frame statistics graphs (Ctrl+F11) and scrolling the global statistics (Ctrl+F2) to the very bottom, and reproing the lag. Take a video of this please.
Another thing, if you've imported your beatmaps from stable, is to try deleting all beatmaps and importing only a few screens worth and seeing if it still happens.

**Reply by nyannoying1337** [reply]
Alright here is the video with all maps https://streamable.com/6upsax and here without maybe 200/250? https://streamable.com/oqdq7b it seems to be better but still seems to struggle a bit

**Reply by bdach** [reply]
`DelayedLoadWrapper` holding up update seems to stronglypoint at the beatmap set carousel backgrounds being the issue indeed, but I'm unsure why that would be the case. Out of curiosity do you have a spinning disk drive, or an SSD? Might point at texture load latency being the issue.

**Reply by integerrr** [reply]
If it's any help mine looks even worse, ~8000 maps on a USB3.0 HDD
[logs.zip](https://github.com/ppy/osu/files/7367286/logs.zip)

https://user-images.githubusercontent.com/7279624/137777198-f78c2141-969e-4b1e-9a30-f7840c0b9e78.mp4

**Reply by bdach** [reply]
@integerrr yours is also showing a `BeatmapLeaderboard` hit and a `SongSelect` hit (almost a full second). If I were to guess what may be the unifying piece behind those three, it'd probably be either I/O or database.

Mind sharing your `client.db`? Either that or I go acquire a huge beatmap collection somehow...

**Reply by integerrr** [reply]
mine is 71MB, how should i send this to you?

**Reply by bdach** [reply]
Oh jeez I expected large but not this large. I think at this point it's probably easier if I fabricate my own by making a test that imports beatmaps in a loop or something. A google drive link or soemething would also work I suppose if you have that.

**Reply by integerrr** [reply]
I managed to get a good enough compression on the file so I've just emailed it to you, hope it helps.

**Reply by nyannoying1337** [reply]
> `DelayedLoadWrapper` holding up update seems to stronglypoint at the beatmap set carousel backgrounds being the issue indeed, but I'm unsure why that would be the case. Out of curiosity do you have a spinning disk drive, or an SSD? Might point at texture load latency being the issue.

I am on a Spinning disk 7200RPM but internal connected not via USB.  

Edit: @bdach I attached my client.db as well just in case. not sure if that would help considering you have his already but might as well https://drive.google.com/file/d/1v7PTqaDtdRHwhpWVgp2YlIS7p49N4x7S/view?usp=sharing

**Reply by bdach** [reply]
I've checked both `client.db` files I had for the purpose of attempting to diagnose this and I can't repro the issue on either, likely because the client DB does not include the actual backgrounds/files for the beatmaps. That would imply that this is not due to database/EF load alone.

I guess I have to look into a method of getting a sufficiently large collection of files to attempt to reproduce.

**Reply by peppy** [reply]
@bdach I'm not sure that's going to help. This is most definitely going to be an IO or GPU stall, which won't be repro'd no matter the size of beatmap collection.

**Reply by nyannoying1337** [reply]
As an update. 

The lag is gone since the Realm update (at least for me ) (2022.127.0)
the only issue that seems to have popped up are little lags when selecting the songs now 

https://user-images.githubusercontent.com/17437302/151391251-6ab4a6fc-df03-4de0-b065-a07ce1c24ac0.mp4

**Reply by peppy** [reply]
Great to hear!
